libursa/OnDiskIndex.cpp

```diff
@@ -281,13 +283,90 @@ std::vector<FileId> OnDiskIndex::query_primitive(TriGram trigram) const {
 
 unsigned long OnDiskIndex::real_size() const { return fs::file_size(fpath); }
 
+// Merge the indexes, and stream the results to the `out` stream immediately.
+// This function tries to batch reads, which makes it much more efficient on
+// HDDs (on SSDs the difference is not noticeable).
+// Instead of reading one ngram run at a time, read up to MAX_BATCH ngrams.
+// In a special case of batch size=1, this is equivalent to the older method.
+void OnDiskIndex::on_disk_merge_core(
+    const std::vector<IndexMergeHelper> &indexes, RawFile *out, Task *task) {
+    // Offsets to every run in the file (including the header).
+    std::vector<uint64_t> offsets(NUM_TRIGRAMS + 1);
+
+    // Current offset in the file (equal to out.ftell()).
+    uint64_t out_offset = 16;
+
+    // Arbitrary number describing how much RAM we want to spend on the run
+    // cache during the batched stream pass.
+    constexpr uint64_t MAX_BATCH_BYTES = 2ULL * 1024ULL * 1024ULL * 1024ULL;
+
+    // Vector used for all merge passes (to avoid unnecessary reallocations).
+    std::vector<uint8_t> batch_vector(MAX_BATCH_BYTES);
```

Comment:
Do we want to fail in case of memory pressure? :thinking:

Reply:
2GB is memory pressure? I forgot how to count that low.

Reply:
Honest answer though: longer term I have this idea of tasks "locking" RAM. This would allow for optimal memory usage and safety. This, and I'll make it configurable with `config set max_batch_bytes` or something.  Before that I try to add some sane defaults (which may or may not actually be sane).